File: app/src/main/java/com/benny/openlauncher/util/AppManager.java
package com.benny.openlauncher.util;

import android.content.Context;
import android.content.Intent;
import android.content.pm.LauncherActivityInfo;
import android.content.pm.LauncherApps;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.pm.ShortcutInfo;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Process;
import android.os.UserHandle;
import android.os.UserManager;
import android.support.annotation.NonNull;

import com.benny.openlauncher.activity.HomeActivity;
import com.benny.openlauncher.interfaces.AppDeleteListener;
import com.benny.openlauncher.interfaces.AppUpdateListener;
import com.benny.openlauncher.model.App;
import com.benny.openlauncher.model.Item;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.text.Collator;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

public class AppManager {
    private static Logger LOG = LoggerFactory.getLogger("AppManager");

    private static AppManager appManager;

    public static AppManager getInstance(Context context) {
        return appManager == null ? (appManager = new AppManager(context)) : appManager;
    }

    private PackageManager _packageManager;
    private List<App> _apps = new ArrayList<>();
    private List<App> _nonFilteredApps = new ArrayList<>();
    public final List<AppUpdateListener> _updateListeners = new ArrayList<>();
    public final List<AppDeleteListener> _deleteListeners = new ArrayList<>();
    public boolean _recreateAfterGettingApps;
    private AsyncTask _task;
    private Context _context;

    public PackageManager getPackageManager() {
        return _packageManager;
    }

    public Context getContext() {
        return _context;
    }

    public AppManager(Context context) {
        _context = context;
        _packageManager = context.getPackageManager();
    }

    public App findApp(Intent intent) {
        if (intent == null || intent.getComponent() == null) return null;

        String packageName = intent.getComponent().getPackageName();
        String className = intent.getComponent().getClassName();
        for (App app : _apps) {
            if (app._className.equals(className) && app._packageName.equals(packageName)) {
                return app;
            }
        }
        return null;
    }

    public List<App> getApps() {
        return _apps;
    }

    public List<App> getNonFilteredApps() {
        return _nonFilteredApps;
    }

    public void init() {
        getAllApps();
    }

    public void getAllApps() {
        if (_task == null || _task.getStatus() == AsyncTask.Status.FINISHED)
            _task = new AsyncGetApps().execute();
        else if (_task.getStatus() == AsyncTask.Status.RUNNING) {
            _task.cancel(false);
            _task = new AsyncGetApps().execute();
        }
    }

    public List<App> getAllApps(Context context, boolean includeHidden) {
        return includeHidden ? getNonFilteredApps() : getApps();
    }

    public App findItemApp(Item item) {
        return findApp(item.getIntent());
    }

    public App createApp(Intent intent) {
        try {
            ResolveInfo info = _packageManager.resolveActivity(intent, 0);
            List<ShortcutInfo> shortcutInfo = Tool.getShortcutInfo(getContext(), intent.getComponent().getPackageName());
            return new App(_packageManager, info, shortcutInfo);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public void onAppUpdated(Context context, Intent intent) {
        getAllApps();
    }

    public void addUpdateListener(AppUpdateListener updateListener) {
        _updateListeners.add(updateListener);
    }

    public void addDeleteListener(AppDeleteListener deleteListener) {
        _deleteListeners.add(deleteListener);
    }

    public void notifyUpdateListeners(@NonNull List<App> apps) {
        Iterator<AppUpdateListener> iter = _updateListeners.iterator();
        while (iter.hasNext()) {
            if (iter.next().onAppUpdated(apps)) {
                iter.remove();
            }
        }
    }

    public void notifyRemoveListeners(@NonNull List<App> apps) {
        Iterator<AppDeleteListener> iter = _deleteListeners.iterator();
        while (iter.hasNext()) {
            if (iter.next().onAppDeleted(apps)) {
                iter.remove();
            }
        }
    }

    private class AsyncGetApps extends AsyncTask {
        private List<App> appsTemp;
        private List<App> nonFilteredAppsTemp;
        private List<App> removedApps;

        @Override
        protected void onPreExecute() {
            appsTemp = new ArrayList<>();
            nonFilteredAppsTemp = new ArrayList<>();
            removedApps = new ArrayList<>();
            super.onPreExecute();
        }

        @Override
        protected void onCancelled() {
            appsTemp = null;
            nonFilteredAppsTemp = null;
            removedApps = new ArrayList<>();
            super.onCancelled();
        }

        @Override
        protected Object doInBackground(Object[] p1) {

            // work profile support
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                LauncherApps launcherApps = (LauncherApps) _context.getSystemService(Context.LAUNCHER_APPS_SERVICE);
                List<UserHandle> profiles = launcherApps.getProfiles();
                for (UserHandle userHandle : profiles) {
                    List<LauncherActivityInfo> apps = launcherApps.getActivityList(null, userHandle);
                    for (LauncherActivityInfo info : apps) {
                        List<ShortcutInfo> shortcutInfo = Tool.getShortcutInfo(getContext(), info.getComponentName().getPackageName());
                        App app = new App(_packageManager, info, shortcutInfo);
                        app._userHandle = userHandle;
                        LOG.debug("adding work profile to non filtered list: {}, {}, {}", app._label, app._packageName, app._className);
                        nonFilteredAppsTemp.add(app);
                    }
                }
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
                UserManager userManager = (UserManager) _context.getSystemService(Context.USER_SERVICE);
                // LauncherApps.getProfiles() is not available for API 25, so just get all associated user profile handlers
                List<UserHandle> profiles = userManager.getUserProfiles();
                LauncherApps launcherApps = (LauncherApps) _context.getSystemService(Context.LAUNCHER_APPS_SERVICE);
                for (UserHandle userHandle : profiles) {
                    List<LauncherActivityInfo> apps = launcherApps.getActivityList(null, userHandle);
                    for (LauncherActivityInfo info : apps) {
                        List<ShortcutInfo> shortcutInfo = Tool.getShortcutInfo(getContext(), info.getComponentName().getPackageName());
                        App app = new App(_packageManager, info, shortcutInfo);
                        app._userHandle = userHandle;
                        LOG.debug("adding work profile to non filtered list: {}, {}, {}", app._label, app._packageName, app._className);
                        nonFilteredAppsTemp.add(app);
                    }
                }
            } else {
                Intent intent = new Intent(Intent.ACTION_MAIN, null);
                intent.addCategory(Intent.CATEGORY_LAUNCHER);
                List<ResolveInfo> activitiesInfo = _packageManager.queryIntentActivities(intent, 0);
                for (ResolveInfo info : activitiesInfo) {
                    App app = new App(_packageManager, info, null);
                    LOG.debug("adding app to non filtered list: {}, {}, {}", app._label,  app._packageName, app._className);
                    nonFilteredAppsTemp.add(app);
                }
            }

            // sort the apps by label here
            Collections.sort(nonFilteredAppsTemp, new Comparator<App>() {
                @Override
                public int compare(App one, App two) {
                    return Collator.getInstance().compare(one._label, two._label);
                }
            });

            List<String> hiddenList = AppSettings.get().getHiddenAppsList();
            if (hiddenList != null) {
                for (int i = 0; i < nonFilteredAppsTemp.size(); i++) {
                    boolean shouldGetAway = false;
                    for (String hidItemRaw : hiddenList) {
                        if ((nonFilteredAppsTemp.get(i).getComponentName()).equals(hidItemRaw)) {
                            shouldGetAway = true;
                            break;
                        }
                    }
                    if (!shouldGetAway) {
                        appsTemp.add(nonFilteredAppsTemp.get(i));
                    }
                }
            } else {
                appsTemp.addAll(nonFilteredAppsTemp);
            }

            removedApps = getRemovedApps(_apps, appsTemp);

            for (App app : removedApps) {
                HomeActivity._db.deleteItems(app);
            }

            AppSettings appSettings = AppSettings.get();
            if (!appSettings.getIconPack().isEmpty() && Tool.isPackageInstalled(appSettings.getIconPack(), _packageManager)) {
                IconPackHelper.applyIconPack(AppManager.this, Tool.dp2px(appSettings.getIconSize()), appSettings.getIconPack(), appsTemp);
            }
            return null;
        }

        @Override
        protected void onPostExecute(Object result) {
            _apps = appsTemp;
            _nonFilteredApps = nonFilteredAppsTemp;

            if (removedApps.size() > 0) {
                notifyRemoveListeners(removedApps);
            }

            notifyUpdateListeners(appsTemp);

            if (_recreateAfterGettingApps) {
                _recreateAfterGettingApps = false;
                if (_context instanceof HomeActivity)
                    ((HomeActivity) _context).recreate();
            }

            super.onPostExecute(result);
        }
    }

    public static List<App> getRemovedApps(List<App> oldApps, List<App> newApps) {
        List<App> removed = new ArrayList<>();
        // if this is the first call then return an empty list
        if (oldApps.size() == 0) {
            return removed;
        }
        for (int i = 0; i < oldApps.size(); i++) {
            if (!newApps.contains(oldApps.get(i))) {
                removed.add(oldApps.get(i));
                break;
            }
        }
        return removed;
    }
}

File: app/src/main/java/com/benny/openlauncher/util/AppManager.kt
package com.benny.openlauncher.util

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.os.AsyncTask
import android.support.v4.app.ActivityCompat
import android.view.Gravity
import com.afollestad.materialdialogs.MaterialDialog
import com.benny.openlauncher.R
import com.benny.openlauncher.activity.Home
import com.benny.openlauncher.core.activity.CoreHome
import com.benny.openlauncher.core.interfaces.AbstractApp
import com.benny.openlauncher.core.interfaces.AppDeleteListener
import com.benny.openlauncher.core.interfaces.AppUpdateListener
import com.benny.openlauncher.core.manager.Setup
import com.benny.openlauncher.core.model.IconLabelItem
import com.benny.openlauncher.core.model.Item
import com.benny.openlauncher.core.util.BaseIconProvider
import com.benny.openlauncher.core.util.Tool
import com.benny.openlauncher.core.util.toPx
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter
import java.text.Collator
import java.util.*

class AppManager(val context: Context) : Setup.AppLoader<AppManager.App> {

    val packageManager: PackageManager = context.packageManager
    val apps = ArrayList<App>()
    val nonFilteredApps = ArrayList<App>()
    private val updateListeners: MutableList<AppUpdateListener<App>> = ArrayList()
    private val deleteListeners: MutableList<AppDeleteListener<App>> = ArrayList()
    var recreateAfterGettingApps: Boolean = false

    private var task: AsyncTask<*, *, *>? = null

    fun findApp(intent: Intent?): App? {
        if (intent == null || intent.component == null) return null

        val packageName = intent.component!!.packageName
        val className = intent.component!!.className
        for (app in apps) {
            if (app.className == className && app.packageName == packageName) {
                return app
            }
        }
        return null
    }

    fun getApps(): List<App> = apps

    fun getNonFilteredApps(): List<App> = nonFilteredApps

    fun clearListener() {
        updateListeners.clear()
        deleteListeners.clear()
    }

    fun init() {
        getAllApps()
    }

    private fun getAllApps() {
        if (task == null || task!!.status == AsyncTask.Status.FINISHED)
            task = AsyncGetApps().execute()
        else if (task!!.status == AsyncTask.Status.RUNNING) {
            task!!.cancel(false)
            task = AsyncGetApps().execute()
        }
    }

    fun startPickIconPackIntent(activity: Activity) {
        val intent = Intent(Intent.ACTION_MAIN)
        intent.addCategory("com.anddoes.launcher.THEME")

        val fastItemAdapter = FastItemAdapter<IconLabelItem>()

        val resolveInfos = packageManager.queryIntentActivities(intent, 0)
        Collections.sort(resolveInfos, ResolveInfo.DisplayNameComparator(packageManager))
        val d = MaterialDialog.Builder(activity)
                .adapter(fastItemAdapter, null)
                .title(activity.getString(R.string.dialog__icon_pack_title))
                .build()

        fastItemAdapter.add(IconLabelItem(activity, R.drawable.ic_launcher, R.string.label_default, -1)
                .withDrawablePadding(8.toPx().toFloat())
                .withIconGravity(Gravity.START)
                .withOnClickListener {
                    recreateAfterGettingApps = true
                    AppSettings.get().iconPack = ""
                    getAllApps()
                    d.dismiss()
                })

        for (i in resolveInfos.indices) {
            fastItemAdapter.add(IconLabelItem(activity, resolveInfos[i].loadIcon(packageManager), " " + resolveInfos[i].loadLabel(packageManager).toString(), -1)
                    .withIconGravity(Gravity.START)
                    .withOnClickListener {
                        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                            recreateAfterGettingApps = true
                            AppSettings.get().iconPack = resolveInfos[i].activityInfo.packageName
                            getAllApps()
                            d.dismiss()
                        } else {
                            Tool.toast(context, activity.getString(R.string.dialog__icon_pack_info_toast))
                            ActivityCompat.requestPermissions(Home.launcher!!, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), CoreHome.REQUEST_PERMISSION_STORAGE)
                        }
                    })
        }
        d.show()
    }

    fun onReceive(p1: Context, p2: Intent) {
        getAllApps()
    }

    // -----------------------
    // AppLoader interface
    // -----------------------

    override fun loadItems() {
        getAllApps()
    }

    override fun getAllApps(context: Context, includeHidden: Boolean): List<App> {
        return if (includeHidden) getNonFilteredApps() else getApps()
    }

    override fun findItemApp(item: Item): App? {
        return findApp(item.intent)
    }

    override fun createApp(intent: Intent): App? {
        try {
            val info = packageManager.resolveActivity(intent, 0)
            val app = App(context, info, packageManager)
            if (apps != null && !apps.contains(app))
                apps.add(app)
            return app
        } catch (e: Exception) {
            return null
        }

    }

    override fun onAppUpdated(p1: Context, p2: Intent) {
        onReceive(p1, p2)
    }

    override fun notifyUpdateListeners(apps: List<App>) {
        val iter = updateListeners.iterator()
        while (iter.hasNext()) {
            if (iter.next().onAppUpdated(apps)) {
                iter.remove()
            }
        }
    }

    override fun notifyRemoveListeners(apps: List<App>) {
        val iter = deleteListeners.iterator()
        while (iter.hasNext()) {
            if (iter.next().onAppDeleted(apps)) {
                iter.remove()
            }
        }
    }

    override fun addUpdateListener(updateListener: AppUpdateListener<AppManager.App>) {
        updateListeners.add(updateListener)
    }

    override fun removeUpdateListener(updateListener: AppUpdateListener<AppManager.App>) {
        updateListeners.remove(updateListener)
    }

    override fun addDeleteListener(deleteListener: AppDeleteListener<AppManager.App>) {
        deleteListeners.add(deleteListener)
    }

    override fun removeDeleteListener(deleteListener: AppDeleteListener<AppManager.App>) {
        deleteListeners.remove(deleteListener)
    }

    private inner class AsyncGetApps : AsyncTask<Void, Void, Void>() {

        private var tempApps: List<App>? = null

        override fun onPreExecute() {
            tempApps = ArrayList(apps)
            super.onPreExecute()
        }

        override fun onCancelled() {
            tempApps = null
            super.onCancelled()
        }

        protected override fun doInBackground(vararg p0: Void?): Void? {
            apps.clear()
            nonFilteredApps.clear()

            val intent = Intent(Intent.ACTION_MAIN, null)
            intent.addCategory(Intent.CATEGORY_LAUNCHER)
            val activitiesInfo = packageManager.queryIntentActivities(intent, 0)
            Collections.sort(activitiesInfo) { p1, p2 -> Collator.getInstance().compare(p1.loadLabel(packageManager).toString(), p2.loadLabel(packageManager).toString()) }

            for (info in activitiesInfo) {
                val app = App(context, info, packageManager)
                nonFilteredApps.add(app)
            }

            val hiddenList = AppSettings.get().hiddenAppsList
            if (hiddenList != null) {
                for (i in nonFilteredApps.indices) {
                    var shouldGetAway = false
                    for (hidItemRaw in hiddenList) {
                        if (nonFilteredApps[i].packageName + "/" + nonFilteredApps[i].className == hidItemRaw) {
                            shouldGetAway = true
                            break
                        }
                    }
                    if (!shouldGetAway) {
                        apps.add(nonFilteredApps[i])
                    }
                }
            } else {
                for (info in activitiesInfo)
                    apps.add(App(context, info, packageManager))
            }

            val appSettings = AppSettings.get()
            if (!appSettings.iconPack.isEmpty() && Tool.isPackageInstalled(appSettings.iconPack, packageManager)) {
                IconPackHelper.themePacs(this@AppManager, Tool.dp2px(appSettings.iconSize, context), appSettings.iconPack, apps)
            }
            return null
        }

        override fun onPostExecute(result: Void?) {

            notifyUpdateListeners(apps)

            val removed = Tool.getRemovedApps(tempApps!!, apps)
            if (removed.size > 0) {
                notifyRemoveListeners(removed)
            }

            if (recreateAfterGettingApps) {
                recreateAfterGettingApps = false
                (context as? Home)?.recreate()
            }

            super.onPostExecute(result)
        }
    }

    data class App(
            val context: Context,
            var info: ResolveInfo,
            val packageManager: PackageManager,
            override var label: String = info.loadLabel(packageManager).toString(),
            override var packageName: String = info.activityInfo.packageName,
            override var className: String = info.activityInfo.name,
            override var iconProvider: BaseIconProvider = Setup.imageLoader().createIconProvider(info.loadIcon(packageManager))
    ) : AbstractApp {

        override fun hashCode(): Int {
            var result = context.hashCode()
            result = 31 * result + info.hashCode()
            result = 31 * result + packageManager.hashCode()
            result = 31 * result + label.hashCode()
            result = 31 * result + packageName.hashCode()
            result = 31 * result + className.hashCode()
            result = 31 * result + iconProvider.hashCode()
            return result
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as App

            if (context != other.context) return false
            if (info != other.info) return false
            if (packageManager != other.packageManager) return false
            if (label != other.label) return false
            if (packageName != other.packageName) return false
            if (className != other.className) return false
            if (iconProvider != other.iconProvider) return false

            return true
        }
    }

    abstract class AppUpdatedListener : AppUpdateListener<App> {
        private val listenerID: String = UUID.randomUUID().toString()

        override fun equals(other: Any?): Boolean = other is AppUpdatedListener && other.listenerID == this.listenerID
    }

    companion object {
        private var ref: AppManager? = null

        fun getInstance(context: Context): AppManager = if (ref == null) {
            ref = AppManager(context)
            ref!!
        } else
            ref!!
    }
}
